### PR TITLE
Wrap state to a box struct for save TypeTag of state and refactor the RawTable

### LINKED ISSUE #65: [statedb] Wrap Object to ValueBox
Part of #21 

Currently, We directly store Object BCS binary to the SMT.

For security, we need to warp the Object to ValueBox, and store the `Object<T>`'s struct tag, and ensure the safety of Object deserialize.

TBD:

Do We also need to keep the codec type in the ValueBox? Such as [BCS](https://github.com/diem/bcs) or [RLP](https://ethereum.org/nl/developers/docs/data-structures-and-encoding/rlp/).

Use case: serialize an Object in Solidity and migrate it to off-chain. [rollup-nft](https://github.com/rooch-network/rollup-nft)

### LINKED ISSUE #112: [Security] Object Storage type check
We need to ensure object_storage::borrow or object_storage::contains to check the value type.

```move

#[test_only]
struct TestObject has key{
    f: u8
}
 #[test_only]
struct TestObject2 has key{
    f: u8
}

#[test(sender=@0x42)]
fun test_object_storage(sender: address) {
    let ctx = moveos_std::tx_context::new_test_context(sender);
    let os = new(&mut ctx);
    let object = object::new(&mut ctx, sender, TestObject{f: 1});
    let object_id = object::id(&object);
    add(&mut os, object);
    assert!(contains<TestObject>(&os, object_id), 1000);
    
    //assert!(!contains<TestObject2>(&os, object_id), 1001);
    
    drop_object_storage(os);
}
```

The second `assert` should be true.

### Ref
1. #65

### PR DESCRIPTION
**Pull request by jolestar**
1. Define `State` struct to wrap the Move State, and Save the TypeTag to storage. resolve #65 
  ```
  pub struct State {
    /// the bytes of state
    pub value: Vec<u8>,
    /// the type of state
    pub value_type: TypeTag,
  }
  ```
2. Check the type when load data from storage, and resolve #112 
3. Refactor the RawTable
   * Use ObjectID as TableHandle
   * Refactor the GlobalValue in Table, GlobalValue does not binding Type when creating.
   * Encapsulate the Box<V> within RowTable implementation, auto box, and unbox when deserializing and serializing.
   * The `raw_table::contains<K>(key: K)`, only cares about the Key type, and does not care about the Value type.
4. Update Move for export some TypeTag function https://github.com/move-language/move/commit/c7d8c2b0cdd06dbd90e0ab306932356620b5648a

### CODE REVIEW

File: moveos/moveos-stdlib/src/natives/moveos_stdlib/raw_table/mod.rs [hunk 1]
```diff
@@ -137,8 +137,113 @@ struct TableData {
...
+        } else {
+            self.value_layout_and_type = Some((value_layout, value_type));
+        }
+        self.box_value.move_to(val)?;
+        Ok(())
+    }
+
+    pub fn borrow_global(&self, expect_value_type: TypeTag) -> PartialVMResult<Value> {
+        let value = self.box_value.borrow_global()?;
+        match &self.value_layout_and_type {
```
Comment (baichuan3) [inline moveos/moveos-stdlib/src/natives/moveos_stdlib/raw_table/mod.rs:193]: maybe can abstract type checking as a separate function

**Review by baichuan3 — APPROVED** [review]
LGTM

File: moveos/moveos-stdlib/src/natives/moveos_stdlib/raw_table/mod.rs (hunk 1 above)
Comment (jolestar) [reply moveos/moveos-stdlib/src/natives/moveos_stdlib/raw_table/mod.rs:193]: I write a TODO.

**Comment by jolestar** [comment]
I will merge this PR first.  @steelgeek091 @wow-sven please help continue to review.

#170 conflict with this PR, @baichuan3 please help to do the merge work.

**Pull request merged**